drivers/rtc/rtc_pcf8523.c

```diff
@@ -0,0 +1,966 @@
+		/* Only change INT1 GPIO if alarm interrupt not enabled */
+		err = pcf8523_int1_enable_unlocked(dev, callback != NULL);
+		if (err != 0) {
+			goto unlock;
+		}
+	}
+
+	err = pcf8523_write_reg8(dev, PCF8523_CONTROL_1, control_1);
+	if (err != 0) {
+		goto unlock;
+	}
+
+unlock:
+	k_mutex_unlock(&data->lock);
+
+	/* Wake up the INT1 thread since the seconds flag may already be set */
+	k_sem_give(&data->int1_sem);
+
+	return err;
+}
+#endif /* PCF8523_INT1_GPIOS_IN_USE && defined(CONFIG_RTC_UPDATE) */
+
+#ifdef CONFIG_RTC_CALIBRATION
+
+/* See PCF8523 data sheet table 29 */
+#if defined(CONFIG_RTC_PCF8523_OFFSET_MODE_SLOW)
+#define PCF8523_OFFSET_PPB_PER_LSB 4340
+#elif defined(CONFIG_RTC_PCF8523_OFFSET_MODE_FAST)
+#define PCF8523_OFFSET_PPB_PER_LSB 4069
+#else
+#error Unsupported offset mode
+#endif
+
+#define PCF8523_OFFSET_PPB_MIN (-64 * PCF8523_OFFSET_PPB_PER_LSB)
+#define PCF8523_OFFSET_PPB_MAX (63 * PCF8523_OFFSET_PPB_PER_LSB)
+
+static int pcf8523_set_calibration(const struct device *dev, int32_t freq_ppb)
+{
+	int32_t period_ppb = freq_ppb * -1;
+	int8_t offset;
+
+	if (period_ppb < PCF8523_OFFSET_PPB_MIN || period_ppb > PCF8523_OFFSET_PPB_MAX) {
+		LOG_WRN("calibration value (%d ppb) out of range", freq_ppb);
+		return -EINVAL;
+	}
+
+	offset = period_ppb / PCF8523_OFFSET_PPB_PER_LSB;
+
+	if (IS_ENABLED(CONFIG_RTC_PCF8523_OFFSET_MODE_FAST)) {
+		offset |= PCF8523_OFFSET_MODE;
+	}
+
+	LOG_DBG("freq_ppb = %d, period_ppb = %d, offset = %d", freq_ppb, period_ppb, offset);
+
+	return pcf8523_write_reg8(dev, PCF8523_OFFSET, offset);
+}
+
+static int pcf8523_get_calibration(const struct device *dev, int32_t *freq_ppb)
+{
+	int32_t period_ppb;
+	int8_t offset;
+	int err;
+
+	err = pcf8523_read_reg8(dev, PCF8523_OFFSET, &offset);
+	if (err != 0) {
+		return err;
+	}
+
+	/* Clear mode bit and sign extend the offset */
+	period_ppb = (offset << 1U) >> 1U;
+
+	period_ppb = period_ppb * PCF8523_OFFSET_PPB_PER_LSB;
+	*freq_ppb = period_ppb * -1;
+
+	LOG_DBG("freq_ppb = %d, period_ppb = %d, offset = %d", *freq_ppb, period_ppb, offset);
+
+	return 0;
+}
+#endif /* CONFIG_RTC_CALIBRATION */
+
+static int pcf8523_init(const struct device *dev)
+{
+	const struct pcf8523_config *config = dev->config;
+	struct pcf8523_data *data = dev->data;
+	uint8_t tmr_clkout_ctrl;
+	uint8_t regs[3];
+	int err;
+
+	k_mutex_init(&data->lock);
+
+	if (!i2c_is_ready_dt(&config->i2c)) {
+		LOG_ERR("I2C bus not ready");
+		return -ENODEV;
+	}
+
+#if PCF8523_INT1_GPIOS_IN_USE
+	k_tid_t tid;
+
+	if (config->int1.port != NULL) {
+		k_sem_init(&data->int1_sem, 0, INT_MAX);
+
+		tid = k_thread_create(&data->int1_thread, data->int1_stack,
+				      K_THREAD_STACK_SIZEOF(data->int1_stack),
+				      (k_thread_entry_t)pcf8523_int1_thread, (void *)dev, NULL,
+				      NULL, CONFIG_RTC_PCF8523_THREAD_PRIO, 0, K_NO_WAIT);
+		k_thread_name_set(tid, "pcf8523");
+
+		if (!gpio_is_ready_dt(&config->int1)) {
+			LOG_ERR("GPIO not ready");
+			return -ENODEV;
+		}
```

Comment:
Should the GPIO be checked if it is ready before creating the thread?

Reply:
Sure, I have changed the order.